// fwjs/application/helpers/manager.js
/**
 * Helper manager. Create context helper.
 */

var manager = function( application ){
    this.app = application;
    this.helpers = {};
};

manager.prototype.ERROR_ADD_NO_NAME = {code:'HMGR_1',message:'Can\'t add helper with empty name'};
manager.prototype.ERROR_ADD_INVALID_PATH = {code:'HMGR_2',message:'Can\'t add helper, path is invalid'};
manager.prototype.ERROR_ADD_INVALID_HELPER = {code:'HMGR_3',message:'Can\'t add helper, helper must be a function or a file path.'};
manager.prototype.ERROR_NOT_EXIST = {code:'HMGR_4',message:'Helper does not exist.'};

manager.prototype.configure = function( helpers, path , next ){
    if( helpers instanceof Object ){        
        for( var name in helpers ){
            this.add(name, helpers[name], path );
        }
    }
    next();
};

manager.prototype.add = function( name, helper, path ){
    if( name ){
        if( typeof(helper) === 'string' ){
            try{
                this.helpers[name] = require(path+helper);
                return;
            }catch( e ){
                console.log( this.ERROR_ADD_INVALID_PATH, {name:name,path:path+helper} );
                this.app.throw( this.ERROR_ADD_INVALID_PATH, {name:name,path:path+helper});
            }
        }
        else if( typeof(helper) === 'function' ){
            this.helpers[name] = helper;
            return;
        }
        this.app.throw( this.ERROR_ADD_INVALID_HELPER,{name:name,path:path,helper:''+helper});
    }
    this.app.throw( this.ERROR_ADD_NO_NAME,{path:path});
    
};


manager.prototype.get = function( name, ctx){
    if( this.helpers[name] ){
        return new (this.helpers[name])( ctx );
    }else
        this.app.throw(this.ERROR_NOT_EXIST, {name:name});
};

manager.prototype.exist = function( name ){
    return this.helpers[name]?true:false;
};

module.exports = manager;
