// fwjs/application/views/manager.js
/*
 * Views manager.
 * Load & manage app views.
 */

var util = require('../../utilities/utilities.js');

var manager = function( application, engine ){
    this.app = application;
    this.engine = engine;
    this.views = {};        
};

manager.prototype.configure = function( views, path , next ){    
    var fns = [];
    
    if( views instanceof Object ){
        for( var name in views ){
            if( views[name] instanceof Object ){
                fns.push([this.add.bind(this),[name,views[name],path]]);
            }
        }
    }
    util.parallel(fns)(next);
};

manager.prototype.add = function( name, view, path, next){    
    if( name && typeof(name)==='string' && view && typeof(view)==='string' ){
        
        
        
        
    
    }
};

manager.prototype.get = function( name, context, datas ){
    if( this.views[name] )
        return new (this.views[name])(context,datas);
};

manager.prototype.remove = function( name ){
    this.views[name] = undefined;
};

manager.prototype.exist = function( name ){
    return this.views[name]?true:false;
};

module.exports = manager;
