// fwjs/routeManager/routeManager.js
/* 
 * Route Manager.
 * Build routes configuration / Route request.
 * 
 * Routes config example:
 * var routes = {
 *      nameRouteA:{
 *          type:'literal',
 *          route:'literal/path',
 *          module:'nameOfCalledModule',
 *          action:'moduleActionToCall'
 *      },
 *      nameRouteB:{
 *          type:'segment',
 *          route:'segment/[:dynamicParam]/path',
 *          module:'nameOfCalledModule',
 *          action:'moduleActionToCall',
 *          constraints:{
 *              dynamicParam:'\\w\*' // Have to match with given Regex.
 *          }
 *      },
 *      nameRouteC:{
 *          type:'section',
 *          route:'section/common/',
 *          module:'nameOfCalledModule',
 *          action:'moduleActionToCall',
 *          childs:{
 *              nameChildRoute:{
 *                  type:'literal',
 *                  route:'add',
 *                  module:'nameOfCalledModule',
 *                  action:'moduleActionToCall'
 *              },
 *              nameChildRouteB:{
 *                  type:'segment',
 *                  route:'get/[:element]',
 *                  module:'nameOfCalledModule',
 *                  action:'moduleActionToCall'
 *              },
 *          }
 *      }
 *  }
 */

var routeManager = function( routes, options ){
    this.routes={};
    this.routers={};
    this.compiledRoutes={};
    this.options = {types:{literal:'literal',section:'section',segment:'segment'/*Route type name: route type path*/}};
    
    if( routes )
        this.setRoutes(routes);
    if( options )
        this.setOptions(options);
};


routeManager.prototype.addRoutes = function(routes){
    if( routes instanceof Object ){
        for( var route in routes )
            this.routes[route] = routes[route];
    }
    return this;
};

routeManager.prototype.setRoutes = function(routes){
    this.routes=routes;
    return this;
};
routeManager.prototype.setOptions = function(options){
    if( options instanceof Object ){
        var keys = Object.keys(options);
        for( var key in keys )
            this.options[key] = options[key];
    }
    return this;
};

routeManager.prototype.configure = function(){
    // Initialize routers objects.
    if( this.options.types )
        for( var type in this.options.types )
            if(!this.routers[type])
                this.routers[type] = new require(this.options.types[type])(this);
    // Generate compiled routes.
    for( var route in this.routes )
        this.routers[this.routes[route].type].compile(this.routes[route],route);
    return this;
};

routeManager.prototype.resolve = function(route){
    var routeObject = false;
    for( var router in this.routers ){
        routeObject = this.routers[router].match(route);
        if( routeObject )
            return routeObject;
    }
    return routeObject;
};

routeManager.prototype.url = function(routeName,parameters){
    for( var router in this.routers ){
        var url = this.routers[router].url(routeName,parameters);
        if( url )
            return url;
    }
};

module.exports = routeManager;
