// fwjs/application/application.js
/* 
 */

var context = require('./context.js'),
    controller = require('./controller.js'),
    jhtml = require('jhtml');
    
var application = function(config){
    //this.util = {url:urlParser};
    this.events = ['handle','route','render','send'];
    this.eventStack={handle:[],route:[],render:[]};
    
    if(config)
        this.configure(config);
};

application.prototype.configure = function(config){
    this.config = config;
    this.setRouteManager(this.config.routes||{});
    this.setModuleManager(this.config.modules||{});
};

application.prototype.setModuleManager = function(configuration){
    if( !this.moduleManager )
        this.moduleManager = new (require('../moduleManager/manager.js'))(this);
    
    for( var module in configuration )
        this.moduleManager.load( module, configuration[module] );
};

application.prototype.setRouteManager = function(routes,options){
    if( !this.routeManager )
        this.routeManager = new (require('../routeManager/manager.js'))(routes,options);
};

//// EVENT MANAGEMENT \\\\

// Bind actions you have to do after event was processed. 
// ( For example, all methods bind on 'route' event, will be processed after application execute route mechanism. )
application.prototype.on = function( event,priority,fn ){
    if( this.eventStack[event] && priority ){
        this.eventStack[event].priorities.push(priority);
        this.eventStack[event].priorities.sort();
        this.eventStack[event].functions[priority] = fn;
    }
};

application.prototype.emit = function(event,context){
    if( this.eventStack[event] ){
        var app=this; var callback=function(){ app.next(event,context); };
        for( var i=this.eventStack[event].priorities.length-1; i>0; i--){
            callback = function(){
                this.eventStack[event].functions[this.eventStack[event].priorities[i]](context,callback); };
        }
        callback();
    }
};

application.prototype.next = function(event,context){
    this[this.events[this.events.indexOf(event)+1]](context);
};

// Applications events methods
application.prototype.handle = function(request,response){
    this.emit('handle',new context(this,request,response));
};

application.prototype.route = function(context){
    
    var routeObject = this.routeManager.resolve( context.request.parsedUrl.pathname );
    if( !routeObject ){
        //Load default config & redirect to 404 page.
        context.response.statusCode = 404;
        
    }else if(routeObject.controller && routeObject.action && this.config.controllers[routeObject.controller] ){
        var ctrl = new (require( this.config.controllers[routeObject.controller] ))(this,context);
        ctrl[routeObject.action]();
    }
    
    
    
};

application.prototype.render = function(context){
    
};

application.prototype.send = function(context){
    
};




module.exports = application;
