// fwjs/events/manager.js
/* 
 * Event Manager.
 * Load event configuration.
 */
var util = require('../utilities/utilities.js');

var manager = function( application ){
    this.app = application;
    this.events = {}; 
};

manager.prototype.configure = function( events, next ){
    if( events instanceof Object ){
        var l, i;
        for( var evt in events ){
            if( Array.isArray(events[evt]) ){
                l = events[evt].length, i=0;
                for(;i<l;i++)
                    this.on(evt,events[evt][i]);
            }
        }
    }
    if( next )
        next();
};

manager.prototype.on = function( event, datas ){
    if( typeof(datas.callback) === 'function' && typeof(event)==='string' ){
        var type = datas.type || 'sync';
        
        if( !this.events[event] )
            this.events[event] = {sync:{},async:[],final:undefined,priorities:[]};
        
        if( type === 'sync'){
            var priority;
            if( datas.priority ){
                priority = datas.priority;
                this.events[event].priorities.push( priority );
                this.events[event].priorities.sort();                
            }else{
                var l = this.events[event].priorities.length;
                priority = l?this.events[event].priorities[l-1]+1:0;
                this.events[event].priorities.push(priority);
            }
            if( this.events[event].sync[priority] )
                this.events[event].sync[priority].push(datas.callback);
            else
                this.events[event].sync[priority] = [datas.callback];
            
        }else if( type === 'async'){
            this.events[event].async.push(datas.callback);
        }else if( type === 'final'){
            this.events[event].final = datas.callback;
        }
    }
    return this;
};

manager.prototype.off = function( event, type, fn, priority ){
    if( typeof(event) ==='string' && this.events[event] ){
        
        if( !fn && !priority && !type )
            this.events[event] = undefined;
        else if( !priority && !fn ){
            if( type === 'sync' ){
                this.events[event].priorities = [];
                this.events[event].sync = {};
            }
            else if( type === 'async')
                this.events[event].async = []; 
            else if( type === 'final')
                this.events[event].final = undefined;
        }
        else{
            
            
            
            
            
        }
    }
    return this;
};

manager.prototype.emit = function( event, args ){
    if( typeof(event) === 'string' && this.events[event] ){
        
        var asyncs=[], i=0, asynclength=this.events[event].async.length,
            syncs = [], k=0, p = this.events[event].priorities, synclength = p.length;
    
        if( asynclength ){
            for(;i<asynclength;i++)
                asyncs.push([this.events[event].async[i].bind(this.app),[args]]);
        }
    
        if( synclength ){
            for(;k<synclength;k++)                
                for( var fn in this.events.sync[p[k]] )
                    syncs.push([this.events[event].sync[p[k]][fn].bind(this.app),[args]]);
            
            asyncs.push([util.chain( syncs )]);
        }
        
        util.parallel( asyncs )( this.events[event].final?this.events[event].final.bind(this,args):undefined );
    }
    return this;
};



























manager.prototype.on = function( event, fn, type, priority ){
    if( fn instanceof Function ){
        if( !this.eventStack[event] )
            this.eventStack[event] = {sync:{},async:[],priorities:[],final:undefined};

        type = type||'sync';
        if( type === 'sync' ){
            if( priority ){
                if( this.eventStack[event].sync[priority] ){
                    this.eventStack[event].priorities.push(priority);
                    this.eventStack[event].priorities.sort();
                }
            }else{
                priority = (this.eventStack[event].priorities.length?
                    this.eventStack[event].priorities[this.eventStack[event].priorities.length-1]+1:500);
                this.eventStack[event].priorities.push(priority);
            }
            if( this.eventStack[event].sync[priority] )
                this.eventStack[event].sync[priority].push(fn);
            else
                this.eventStack[event].sync[priority] = [fn];
            
        }else if( type === 'async' ){
            this.eventStack[event].async.push(fn);
        }else if( type === 'final' ){
            this.eventStack[event].final = fn;
        }
    }
    return this;
};

manager.prototype.off = function( event, fn ){
    var evts = (event === '*' || !event )?Object.keys(this.eventStack):event.split(',');
    
    if( fn && fn instanceof Function ){
        for( var i=0; i<evts.length; i++){
            if( this.eventStack[evts[i]] ){                
                // Remove function from async event stack.
                var idx = this.eventStack[evts[i]].async.indexOf(fn);
                if( idx > -1 )
                    this.eventStack[evts[i]].async.splice(idx,1);
                
                // Remove function from sync event stack.
                for( var p in this.eventStack[evts[i]].sync ){
                    idx = this.eventStack[evts[i]].sync[p].indexOf(fn);
                    if( idx > -1 )
                        this.eventStack[evts[i]].sync[p].splice(idx,1);
                    if( !this.eventStack[evts[i]].sync[p].length ){
                        this.eventStack[evts[i]].sync[p] = null;
                        this.eventStack[evts[i]].priorities.splice( this.eventStack[evts[i]].priorities.indexOf(p),1 );
                    }
                }
            }
        }
    }else{
        for( var i=0; i<evts.length; i++)
            if( this.eventStack[evts[i]] )
                this.eventStack[evts[i]] = {sync:{},async:[],priorities:[],final:this.eventStack[evts[i]].final};
    }    
    return this;
};

manager.prototype.emit = function(event, args){
    if( this.eventStack[event] ){
        for( var i=0; i<this.eventStack[event].async.length; i++)
            this.eventStack[event].async[i]( args );
        
        var final = function(){
            if(this.eventStack[event].final)
                this.eventStack[event].final(args);
        }.bind(this);
        
        if( this.eventStack[event].priorities.length ){
            for( i=0; i<this.eventStack[event].priorities.length; i++ ){
                for( var k=0; k<this.eventStack[event].sync[this.eventStack[event].priorities[i]].length; k++ ){
                    var b = final;
                    final = (function(b,i,k){ return function(){ this.eventStack[event].sync[this.eventStack[event].priorities[i]][k]( b, args); }.bind(this);}).bind(this)(b,i,k);
                }   
            }
        }
        final();
    }
};

module.exports = manager;
